wgpu-hal/src/metal/layer_observer.rs

```diff
@@ -0,0 +1,190 @@
+
+        decl.register()
+    })
+}
+
+/// The unique context pointer for this class.
+fn context_ptr() -> *mut c_void {
+    let ptr: *const Class = class();
+    ptr.cast_mut().cast()
+}
+
+/// Same as <https://docs.rs/raw-window-metal/1.1.0/src/raw_window_metal/observer.rs.html#74-132>.
+extern "C" fn observe_value(
+    this: &Object,
+    _cmd: Sel,
+    key_path: *mut Object,
+    object: *mut Object,
+    change: *mut Object,
+    context: *mut c_void,
+) {
+    // An unrecognized context must belong to the super class.
+    if context != context_ptr() {
+        // SAFETY: The signature is correct, and it's safe to forward to
+        // the superclass' method when we're overriding the method.
+        return unsafe {
+            msg_send![
+                super(this, class!(CAMetalLayer)),
+                observeValueForKeyPath: key_path
+                ofObject: object
+                change: change
+                context: context
+            ]
+        };
+    }
+
+    assert!(!change.is_null());
+
+    let key = unsafe { NSKeyValueChangeNewKey };
+    let new: *mut Object = unsafe { msg_send![change, objectForKey: key] };
+    assert!(!new.is_null());
+
+    let to_compare: *const Object =
+        unsafe { msg_send![class!(NSString), stringWithUTF8String: CONTENTS_SCALE.as_ptr()] };
+    let is_equal: BOOL = unsafe { msg_send![key_path, isEqual: to_compare] };
+    if is_equal != NO {
+        // `contentsScale` is a CGFloat, and so the observed value is always a NSNumber.
+        let scale_factor: CGFloat = if cfg!(target_pointer_width = "64") {
+            unsafe { msg_send![new, doubleValue] }
+        } else {
+            unsafe { msg_send![new, floatValue] }
+        };
+
+        // Set the scale factor of the layer to match the root layer.
+        let _: () = unsafe { msg_send![this, setContentsScale: scale_factor] };
+        return;
+    }
+
+    let to_compare: *const Object =
+        unsafe { msg_send![class!(NSString), stringWithUTF8String: BOUNDS.as_ptr()] };
+    let is_equal: BOOL = unsafe { msg_send![key_path, isEqual: to_compare] };
+    if is_equal != NO {
+        // `bounds` is a CGRect, and so the observed value is always a NSNumber.
+        let bounds: CGRect = unsafe { msg_send![new, rectValue] };
+
+        // Set `bounds` and `position` to match the root layer.
+        //
+        // This differs from just setting the `bounds`, as it also takes into account any
+        // translation that the superlayer may have that we'd want to preserve.
+        let _: () = unsafe { msg_send![this, setFrame: bounds] };
+        return;
+    }
+
+    panic!("unknown observed keypath {key_path:?}");
+}
+
+extern "C" fn dealloc(this: &Object, _cmd: Sel) {
+    // Load the root layer if it still exists, and deregister the observer.
+    //
+    // This is not entirely sound, as the ObserverLayer _could_ have been
+    // moved to another layer; but Wgpu does do that, so it should be fine.
```

Comment:
question: Should this say: "Wgpu [sic] _doesn't_ do that"?

Reply:
Right, of course it should ;)